File: uishell/src/main/scala/net/kemitix/thorp/uishell/UploadProgressEvent.scala
package net.kemitix.thorp.uishell

sealed trait UploadProgressEvent {
  def name: String
}

object UploadProgressEvent {

  final case class TransferEvent(
      name: String
  ) extends UploadProgressEvent

  final case class RequestEvent(
      name: String,
      bytes: Long,
      transferred: Long
  ) extends UploadProgressEvent

  final case class ByteTransferEvent(
      name: String
  ) extends UploadProgressEvent

}

File: uishell/src/main/java/net/kemitix/thorp/uishell/UploadProgressEvent.java
package net.kemitix.thorp.uishell;

import lombok.AccessLevel;
import lombok.RequiredArgsConstructor;

public interface UploadProgressEvent {

    static UploadProgressEvent transferEvent(String name) {
        return new TransferEvent(name);
    }
    @RequiredArgsConstructor(access = AccessLevel.PRIVATE)
    class TransferEvent implements UploadProgressEvent {
        public final String name;
    }

    static UploadProgressEvent requestEvent(String name,
                                            long bytes,
                                            long transferred) {
        return new RequestEvent(name, bytes, transferred);
    }
    @RequiredArgsConstructor(access = AccessLevel.PRIVATE)
    class RequestEvent implements UploadProgressEvent {
        public final String name;
        public final long bytes;
        public final long transferred;
    }

    static UploadProgressEvent bytesTransferEvent(String name) {
        return new TransferEvent(name);
    }
    @RequiredArgsConstructor(access = AccessLevel.PRIVATE)
    class ByteTransferEvent implements UploadProgressEvent {
        public final String name;
    }
}
